api/api.ts

```diff
@@ -1,8 +1,90 @@
-import axios from 'axios';
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import axios, { AxiosError, AxiosPromise, AxiosResponse } from 'axios';
+import { ApiPromise, ApiResponse, EmptyPayload, StatusMessageType } from '../types/apiResponse';
 
-const api = axios.create({
-  baseURL: '/api/',
-  timeout: 10 * 1000, // 10 seconds
-});
+class BaseApi {
+  client = axios.create({
+    baseURL: '/api/',
+    timeout: 10 * 1000, // 10 seconds
+  });
+
+  private getData<D>(url: string, params?: any): AxiosPromise<ApiResponse<D>> {
+    return this.client.get(url, params);
+  }
+
+  private postData<D>(url: string, data: any = {}): AxiosPromise<ApiResponse<D>> {
+    return this.client.post(url, data);
+  }
+
+  private patchData<D>(url: string, data: any = {}): AxiosPromise<ApiResponse<D>> {
+    return this.client.patch(url, data);
+  }
+
+  private deleteData<D>(url: string): AxiosPromise<ApiResponse<D>> {
+    return this.client.delete(url);
+  }
+
+  public get<D>(url: string, params?: any): ApiPromise<D> {
+    return processRequest(url, this.getData(url, params));
+  }
+
+  public post<D>(url: string, data: any = {}): ApiPromise<D> {
+    return processRequest(url, this.postData(url, data));
+  }
+
+  public patch<D>(url: string, data: any = {}): ApiPromise<D> {
+    return processRequest(url, this.patchData(url, data));
+  }
+
+  public delete<D>(url: string): ApiPromise<D> {
+    return processRequest(url, this.deleteData(url));
+  }
+}
+
+function processRequest<D>(endpoint: string, promise: AxiosPromise<ApiResponse<D>>): ApiPromise<D> {
+  return promise
+    .then((response: AxiosResponse) => {
```

Comment:
```suggestion
    .then((response: AxiosResponse<ApiResponse<D>>) => {
```

Not required to compile, but I think it would still be good to add in case we need more type information about `apiResponse` in future. Currently `apiResponse` is inferred to be `any`, but wit this change it is now inferred as `ApiResponse<D>`.